### PR TITLE
Add uptime-kuma extension

### PR DESCRIPTION
**Pull request by MarlBurroW**
## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

### CODE REVIEW

**Comment by MarlBurroW** [comment]
Reviewer(s):

If you don't have any UptimeKuma server available, I guess you can test the extension with a demo server provided by the Uptime Kuma repo: https://demo.uptime.kuma.pet/

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
I just tried the demo and created an API key but I just got this error

uptime-kuma 2023-10-17 at 11 53 59

image

Does the API work with the demo account, can't we use the API key from inside kuma?

**Comment by MarlBurroW** [comment]
Here's your translation to English:

Hello, thank you for your test.
Sorry, indeed the test on a demo instance does not work because there is likely a sticky session mechanism at the proxy level.
So, I've created a test instance for you.

URL: https://kumademo.marlburrow.io
USERNAME: kumademo
PASSWORD: kumademo

I've added some monitors, but feel free to add more or play around with this instance, I will delete it once the review is finished.

To answer your question, the API Key that can be generated from the settings does not allow access to the socket.
And for my plugin, I prefer to use the socket as it provides information very quickly without having to do polling every x amount of time.
It's a somewhat "hacky" solution, but it works very well; I already use it on another plugin: https://github.com/MarlBurroW/Streamdeck-Uptime-Kuma.

**Comment by pernielsentikaer** [comment]
That was better @MarlBurroW 🔥 

A single question: 
- Is there a way to use the API instead of getting the token from the browser?

Don't change the navigationTitle in the root command - it will be automatically set to the command name. Use navigationTitle only in nested screens to provide additional context 🙂

**Comment by MarlBurroW** [comment]
Ok I agree, this is not a good user experience to have to get an auth token from the browser, you're right.
I made this because the auth flow is more complex with 2 Factors Authentication (I can't manage the 2FA correctly with the preference API provided by raycast).
With some extra works, I've build my own login form to handle the authentication step.
This custom form allow me this things:
- User credentials are not saved in the preferences.
- Only the auth token is stored with LocalStorage API.
- I can add custom validation rules on form fields (like check the socket connection when the user enter the URL field)

I've also improved the UI a bit.

Hope you'll found this authentication flow better than the previous hacky JS console flow.

**Comment by pernielsentikaer** [comment]
Can I push some suggestions @MarlBurroW

File: extensions/uptime-kuma/src/auth-form.tsx [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        error={codeError}
+        id="kuna_2fa"
+        title="2FA Code"
+        defaultValue=""
+      />
+    </Form>
+  );
+}
+
+export default AuthForm;
```
Comment (pernielsentikaer) [inline extensions/uptime-kuma/src/auth-form.tsx:167]:
```suggestion
import {
  Form,
  ActionPanel,
  Action,
  Icon,
  Toast,
  showToast,
  LocalStorage,
  useNavigation,
  popToRoot,
  launchCommand,
  LaunchType,
} from "@raycast/api";
import { UptimeKuma } from "./modules/UptimeKuma";
import { useState } from "react";
import ListMonitors from "./list-monitors";

export function AuthForm() {
  const [urlError, setUrlError] = useState<string>("");
  const [usernameError, setUsernameError] = useState<string>("");
  const [passwordError, setPasswordError] = useState<string>("");
  const [codeError, setCodeError] = useState<string>("");
  const [loading, setLoading] = useState<boolean>(false);

  const { push } = useNavigation();

  async function handleSubmit(values: {
    kuma_url: string;
    kuma_username: string;
    kuma_password: string;
    kuna_2fa: string;
  }) {
    setLoading(true);

    try {
      await checkUrl(values.kuma_url);
    } catch (error) {
      showToast({
        title: "Unable to connect to socket, check your url",
        style: Toast.Style.Failure,
      });
      setLoading(false);
      return;
    }

    let token = "";
    try {
      token = await getToken(values.kuma_url, values.kuma_username, values.kuma_password, values.kuna_2fa);
    } catch (error) {
      showToast({
        title: "Unable to get token, please check your credentials",
        style: Toast.Style.Failure,
      });
      setLoading(false);
      return;
    }

    await LocalStorage.setItem("kuma_url", values.kuma_url);
    await LocalStorage.setItem("kuma_token", token);

    await popToRoot();
    await showToast({
      title: "Login Successful",
      message: "Please open the command again",
    });
  }

  function getToken(url: string, username: string, password: string, code: string): Promise<string> {
    return new Promise((resolve, reject) => {
      const kuma = new UptimeKuma(url);

      kuma.on("connected", () => {
        kuma.getToken(username, password, code);
      });

      kuma.on("token", (token) => {
        kuma.disconnect();
        resolve(token);
      });

      kuma.on("error", (error) => {
        kuma.disconnect();
        reject(error);
      });

      kuma.connect();
    });
  }

  function checkUrl(url: string) {
    return new Promise((resolve, reject) => {
      const checker = new UptimeKuma(url);

      checker.on("connected", () => {
        resolve(true);
        checker.disconnect();
      });

      checker.on("error", (error) => {
        reject(error);
        checker.disconnect();
      });

      checker.connect();
    });
  }

  return (
    <Form
      isLoading={loading}
      actions={
        <ActionPanel>
          <Action.SubmitForm icon={Icon.Upload} title="Login" onSubmit={handleSubmit} />
        </ActionPanel>
      }
    >
      <Form.TextField
        placeholder={"https://uptime.example.com"}
        id="kuma_url"
        title="Kuma url"
        error={urlError}
        defaultValue=""
        onBlur={(event) => {
          if (event.target.value) {
            checkUrl(event.target.value)
              .then(() => {
                setUrlError("");
              })
              .catch((error) => {
                setUrlError(error);
              });
          } else {
            setUrlError("Please enter a valid url");
          }
        }}
      />
      <Form.TextField
        placeholder={"Your kuma username"}
        error={usernameError}
        id="kuma_username"
        title="Kuma Username"
        defaultValue=""
        onBlur={(event) => {
          if (event.target.value) {
            setUsernameError("");
          } else {
            setUsernameError("Please enter a username");
          }
        }}
      />
      <Form.PasswordField
        placeholder={"Your kuma password"}
        error={passwordError}
        id="kuma_password"
        title="Kuma Password"
        defaultValue=""
        onBlur={(event) => {
          if (event.target.value) {
            setPasswordError("");
          } else {
            setPasswordError("Please enter a password");
          }
        }}
      />
      <Form.TextField
        placeholder={"Code from your authenticator app (if 2FA enabled)"}
        error={codeError}
        id="kuna_2fa"
        title="2FA Code"
        defaultValue=""
      />
    </Form>
  );
}

export default AuthForm;

```

File: extensions/uptime-kuma/src/list-monitors.tsx [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+                onAction={logout}
+              />
+            </ActionPanel>
+          }
+          icon={{ source: getMonitorStatusIcon(monitor), tintColor: getMonitorStatusColor(monitor) }}
+        />
+      ))}
+    </List>
+  );
+}
```
Comment (pernielsentikaer) [inline extensions/uptime-kuma/src/list-monitors.tsx:187]:
```suggestion
import {
  Action,
  ActionPanel,
  Alert,
  Color,
  confirmAlert,
  Icon,
  List,
  LocalStorage,
  popToRoot,
  showToast,
  Toast,
  useNavigation,
} from "@raycast/api";
import { AvgPing, Heartbeat, HeartbeatList, Monitor, Uptime, UptimeKuma } from "./modules/UptimeKuma";
import { useEffect, useState } from "react";
import { MonitorDetail } from "./monitor-detail";
import { getMonitorStatusColor, getMonitorStatusIcon, getAccessories } from "./utils/display";
import { useAppStore } from "./utils/store";
import { AuthForm } from "./auth-form";
import { ErrorScreen } from "./error-screen";

export default function Command() {
  const [kuma_url, setKumaUrl] = useState<string>("");
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const [kumaInstance, setKumaInstance] = useState<UptimeKuma | null>(null);
  const {
    monitors,
    setMonitors,
    updateMonitorHeartbeat,
    updateMonitorUptime,
    updateMonitorAvgPing,
    updateMonitorHeartbeatList,
  } = useAppStore();

  const { push } = useNavigation();

  async function logout() {
    await confirmAlert({
      title: "Logout",
      icon: { source: Icon.Power, tintColor: Color.Red },
      message: "Are you sure you want to logout?",
      primaryAction: {
        title: "Log out",
        style: Alert.ActionStyle.Destructive,
        onAction: async () => {
          await popToRoot();
          LocalStorage.removeItem("kuma_url");
          LocalStorage.removeItem("kuma_token");
          setMonitors([]);
          await showToast({ title: "Success", message: "Successfully logged out" });
        },
      },
    });
  }

  async function initCommand() {
    const kuma_url: string | undefined = await LocalStorage.getItem("kuma_url");
    const kuma_token: string | undefined = await LocalStorage.getItem("kuma_token");

    if (!kuma_url || !kuma_token) {
      setIsLoading(false);
      return;
    }

    setIsLoading(true);

    setKumaUrl(kuma_url);

    const kuma = new UptimeKuma(kuma_url, kuma_token);

    setKumaInstance(kuma);

    kuma.on("connected", () => {
      showToast({
        title: "Socket Connected",
        style: Toast.Style.Success,
      });
      kuma.authenticate();
    });

    kuma.on("error", (error) => {
      showToast({
        title: "Error",
        message: error,
        style: Toast.Style.Failure,
      });

      setIsLoading(false);
      kuma.disconnect();

      const errorMessage = `
**Error**: ${error}

If you think your connection settings are wrong, press Enter to run the Login Action and enter your new settings.
`;

      push(<ErrorScreen title={"Socket error"} message={errorMessage} />);
    });

    kuma.on("monitorList", (newMonitors) => {
      // Convert the received object to an array
      const updatedMonitors = Object.values(newMonitors) as Array<Monitor>;

      setMonitors(updatedMonitors);

      setIsLoading(false);
    });

    kuma.on("heartbeatList", (payload: HeartbeatList) => {
      updateMonitorHeartbeatList(payload);
    });

    kuma.on("heartbeat", (hearbeat: Heartbeat) => {
      updateMonitorHeartbeat(hearbeat);
    });

    kuma.on("avgPing", (avgPing: AvgPing) => {
      updateMonitorAvgPing(avgPing);
    });

    kuma.on("uptime", (payload: Uptime) => {
      updateMonitorUptime(payload);
    });

    kuma.connect();
  }

  useEffect(() => {
    initCommand();
  }, []);

  console.log(isLoading);

  return (
    <List isLoading={isLoading}>
      {kuma_url.length === 0 && (
        <List.EmptyView
          title="Please Login"
          icon={Icon.Warning}
          description="You need to login before you can use the extension"
          actions={
            <ActionPanel>
              <Action.Push title="Login" icon={Icon.Power} target={<AuthForm />} />
            </ActionPanel>
          }
        />
      )}
      {monitors.map((monitor) => (
        <List.Item
          id={monitor.id.toString()}
          key={monitor.id}
          title={monitor.name}
          accessories={getAccessories(monitor)}
          actions={
            <ActionPanel>
              <Action.Push title="Monitor Detail" icon={Icon.Info} target={<MonitorDetail monitorId={monitor.id} />} />
              <Action
                title={`${monitor.active ? "Pause Monitor" : "Resume Monitor"}`}
                onAction={() => {
                  if (monitor.active) {
                    kumaInstance?.pauseMonitor(monitor.id, () => {
                      showToast({
                        title: "Monitor paused",
                        style: Toast.Style.Success,
                      });
                    });
                  } else {
                    kumaInstance?.resumeMonitor(monitor.id, () => {
                      showToast({
                        title: "Monitor resumed",
                        style: Toast.Style.Success,
                      });
                    });
                  }
                }}
                icon={
                  monitor.active
                    ? { source: Icon.Pause, tintColor: Color.Yellow }
                    : { source: Icon.Play, tintColor: Color.Green }
                }
              />

              <Action.OpenInBrowser
                title={"Open Monitor In Kuma Dashboard"}
                url={`${kuma_url}/dashboard/${monitor.id}`}
              />
              {monitor.type == "http" && (
                <Action.OpenInBrowser title={"Open Target In Browser"} url={`${monitor.url}`} />
              )}

              <Action
                title={"Logout"}
                icon={{
                  source: Icon.Power,
                  tintColor: Color.Red,
                }}
                style={Action.Style.Destructive}
                onAction={async () => await logout()}
              />
            </ActionPanel>
          }
          icon={{ source: getMonitorStatusIcon(monitor), tintColor: getMonitorStatusColor(monitor) }}
        />
      ))}
    </List>
  );
}

```

**Review by pernielsentikaer — COMMENTED** [review]
I think I would do it like this instead until we have support for a <push> that goes back to root if escape is pressed, right now it only goes back to the view before pusing

**Comment by MarlBurroW** [comment]
Cool ! was'nt aware about `popToRoot()` function. 
I was also disturbed by this navigation behavior, this is a nice patch.

**Comment by pernielsentikaer** [comment]
We probably soon add a way to replace the view instead of just pushing, that gives some new possibilities 🙌

**Comment by MarlBurroW** [comment]
Just pulled your patch and re-applied fix-lint

**Comment by pernielsentikaer** [comment]
Do you mind spinning the demo up again, I'm rewriting a bit more 🙂

**Comment by MarlBurroW** [comment]
Oups yes sorry ! I've rebooted my NAS and forgot to restart the container.
Demo is up and running now.

**Comment by pernielsentikaer** [comment]
Thanks, almost done with the rewrite 🫥

**Comment by pernielsentikaer** [comment]
@MarlBurroW could you check my latest commit 🙂

**Comment by MarlBurroW** [comment]
I reviewed your changes and made some modifications. I also fixed the CI errors, but when I was about to publish, I saw that you had done it too, so I resolved the conflict.

I really like the new error page, it's much nicer than mine. It's more in the Raycast style.
I also like the conditional rendering between the authentication page and the monitors page. It makes navigation cleaner. To make it clearer, I renamed the conditional page to index.ts.
I also updated the screenshots.

**Comment by pernielsentikaer** [comment]
I'm so glad that you the rewrite, it should be a lot more responsive all the way around 🔥

Thanks for letting me help out to make it even more awesome

**Comment by pernielsentikaer** [comment]
Added confirm on logout and removed one unused file

**Review by pernielsentikaer — APPROVED** [review]
Hi 👋 

Looks good to me, approved 🔥

Thanks for your corporation, it was a pleasure working with you 🚀

**Comment by MarlBurroW** [comment]
Thanks @pernielsentikaer , was a pleasure for me too

**Pull request merged**